### PR TITLE
Fix numerical instability in vector_norm when receiving large size tensor

### PR DESCRIPTION
**Pull request by maybeLee**
I made this pull request because I encountered the numerical instability of the API `torch.linalg.vector_norm` when I used it to process a tensor with large shape such as `((1, 32, 224, 224, 160))`.

Here is the code to reproduce the issue:

```
import numpy as np
import torch
x1 = np.ones((1, 32, 224, 224, 160))
ord = 2
print(np.size(x1))  # 256901120
res1 = torch.linalg.vector_norm(torch.tensor(x1, dtype=torch.float32), ord=ord)
res2 = torch.linalg.vector_norm(torch.tensor(x1, dtype=torch.float64), ord=ord)

print(res1, res2)  # tensor(11585.2373) tensor(16028.1353, dtype=torch.float64)
print(f"Expected result: {np.sqrt(np.size(x1))}")  # 16028.135262718493
```

When checking the code, I find that the issue may lie here https://github.com/pytorch/pytorch/blob/3d20cc1332354f5a41cb77c71b0b17cda76e194b/aten/src/ATen/native/cpu/ReduceOpsKernel.cpp#L181

Where precision loss occurs during the accumulation of `acc_vec` when the `acc_vec` is too large while `data_vec * data_vec` is relatively small.

The fix I am applying is to use the Kahan summation algorithm (https://en.wikipedia.org/wiki/Kahan_summation_algorithm).

Please note that after this fix is applied, the average execution time is changed from `0.08817670345306397` to `0.15041379928588866`.
<details>
    <summary>Detailed Code</summary>
import numpy as np
import time
import torch

x1 = np.ones((1, 32, 224, 224, 160))

ord = 2
time_list = []
for i in range(10):
    s_time = time.time()
    res1 = torch.linalg.vector_norm(torch.tensor(x1, dtype=torch.float32), ord=ord)
    time_list.append(time.time() - s_time)
print(np.mean(time_list), np.std(time_list))
</details>

Please let me know if the fix can be improved.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

**Comment by CaoE** [comment]
@maybeLee Thanks for your fix. If we perform group reduction on such large size cases, accuracy should be improved without introducing too much overhead.
For example (pseudocode):
```
acc_t acc_buffer[group_size]:
for (int64_t g = 0; g <  group_size; g++) {
  // Do reduction for each group
  acc_buffer[g] = group_reduce(...)

}

// Do reduction finally
double acc_value = reduce_finaliy(acc_buffer)
result = scalar_t(std::sqrt(acc_value);
```

**Comment by CaoE** [comment]
@maybeLee This is also a group reduce, and it is divided by vec size. But even if the vec size is 32, the size of each group is also large: 256901120/32=8028160. We can set a parameter, e.g.,  group_size = 32768, and further group the reduction externally by group_size (32768 elements are reduced in the current way in each group).
